app.py

```diff
@@ -178,7 +203,46 @@ def azul_match_alias(index, alias, from_=None, size=10):
     dsl = {'term': {'aliases.keyword': alias}}
     if from_:
         dsl['from'] = from_
-    return es_query(index=index, query=dsl, size=size)
+    try:
+        query = es_query(index=index, query=dsl, size=size)
+    # es_query will raise a RuntimeError if it doesn't understand the ES response
+    except RuntimeError:
+        raise ChaliceViewError("Received an unexpected response from Azul.")
+    return query
+
+
+def azul_get_document(key, val, name, es_index, map_fn):
+    """
+    Queries ElasticSearch for a single document and returns a
+    :class:`~chalice.Response` object with the retrieved data. Wrapper
+    around :func:`azul_match_field`. Implements lookup functionality used
+    in :func:`get_data_object` and :func:`get_data_bundle`.
+    :param str key: the key to search for in the given ElasticSearch index
+    :param str val: the value to search for in the given ElasticSearch index
+    :param str name: the key the document should be returned under
+    :param str es_index: the name of the index to query in ElasticSearch
+    :param callable map_fn: function mapping the returned Azul document to a
+                            DOS format
+    :raises RuntimeError: if the ElasticSearch response is not understood
+    :rvtype: :class:`chalice.Response`
+    :returns: the retrieved data or the error state
+    """
+    try:
+        data = azul_match_field(index=es_index, key=key, val=val)
+        data = map_fn(data)
+        # Double check to verify identity
+        if data['id'] != val:
+            raise LookupError("ID mismatch in results")
+    # azul_match_field will also raise a LookupError if no results are returned
+    except LookupError:
+        raise NotFoundError("No results found for type %s and ID %s." % (name, val))
+    # es_query will raise a RuntimeError if it doesn't understand the ES response
+    except RuntimeError:
+        raise ChaliceViewError("Received an unexpected response from Azul.")
+    # if anything else happens...
+    except Exception:
```

Comment:
I think logging should occur in these three cases as well, and that doesn't appear to be the case now.

It short, what I am looking for is anytime an unexpected exception occurs it should be logged, and generally with a stack trace and enough contextual information to identify what caused the problem.
Also, in general, a given exception should only be logged once. That can happen at the "leaf level" functions if a stack trace is provided, or top-level exception catcher if the exceptions are chained as they bubble up.
It depends on what your exception handling and logging strategy is, and currently that's not completely clear to me. The strategy you choose would be worth noting in a comment at the top of the file, so hopefully people working on this code in the future will understand and adhere to the strategy you choose.
Thanks!

Reply:
This makes sense. I'll make these changes and also clarify inline.